Add AWS SageMaker ASR Examples

SeanNaren opened the pull request:
# What does this PR do ?

Adds AWS SageMaker example notebooks for ASR!

**Collection**: ASR

# Changelog 
- Add AWS SageMaker examples to NeMo repo.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

titu1994 commented on nemo/utils/notebook_utils.py:
```diff
@@ -0,0 +1,94 @@
+import tarfile
+from typing import Optional
+
+import librosa
```
Nemo utils should not import librosa, it's not part of core installs. Move the utils up to common or asr collection

SeanNaren replied:
I've changed to importing `sox` and locally importing rather than at the file level! Let me know if thats a reasonable solution for keeping it within utils.

Personally I think having this function is helpful, and could potentially clean up other notebooks that duplicate the same processing. There might be a better place for it to live though, let me know if that's the case

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "outputs": [],
+   "source": [
+    "config_path = str(config_dir / \"config.yaml\")\n",
+    "wget.download(\"https://raw.githubusercontent.com/andrusenkoau/NeMo/esperanto_example/docs/source/asr/examples/esperanto_asr/scripts/get_commonvoice_data_v2.py\", str(code_dir))\n",
```
It's downloading from a user. How about putting this script somewhere Nemo repo ? Or merge it with NeMo CV script

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "source": [
+    "config_path = str(config_dir / \"config.yaml\")\n",
+    "wget.download(\"https://raw.githubusercontent.com/andrusenkoau/NeMo/esperanto_example/docs/source/asr/examples/esperanto_asr/scripts/get_commonvoice_data_v2.py\", str(code_dir))\n",
+    "wget.download(\"https://raw.githubusercontent.com/NVIDIA/NeMo/v1.13.0/examples/asr/conf/conformer/conformer_ctc_bpe.yaml\", config_path)\n",
```
On main, all links point to main branch not r1.x

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!python ./code/get_commonvoice_data_v2.py \\\n",
```
Probably better to go with the MCV + HF sanctioned route here than the internal script that directly access the data. MCV does not like that. It will require the user providing their HF credentials

titu1994 replied:
We have a script for downloading data from HF and it creates a temp manifest which we then update to have NeMo required keys (audio_filepath, text, duration)

titu1994 replied:
@vsl9 please advice

SeanNaren replied:
Spoke to @vsl9 offline, we should not be using the internal script, will move towards the HF approach!

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+    "  --data_root=./data/tokenizers \\\n",
+    "  --tokenizer=\"spe\" \\\n",
+    "  --spe_type=bpe"
+   ]
```
Also add --no_lower_case https://github.com/NVIDIA/NeMo/blob/main/scripts/tokenizers/process_asr_text_tokenizer.py#L48

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+    "\n",
+    "# optimization setup\n",
+    "conf.model.optim.lr = 1e-3\n",
+    "conf.model.optim.sched.name = \"CosineAnnealing\"\n",
```
If you're using cosine annealing, you need to pop out the "d_model" from the sched config.

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+    "conf.exp_manager.exp_dir = \"/opt/ml/model/\"\n",
+    "conf.exp_manager.wandb_logger_kwargs.name = \"cv_sagemaker_8gpus\"\n",
+    "conf.exp_manager.wandb_logger_kwargs.project = \"sagemaker\"\n",
+    "conf.exp_manager.wandb_logger_kwargs.entity = 'nvidia'\n",
```
Entity as Nvidia is for our team, probably remove it and add a comment to put user value?

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+    "\n",
+    "# training setup\n",
+    "conf.trainer.accelerator = \"gpu\"\n",
+    "conf.trainer.max_epochs = 1\n",
```
Max epochs 1?

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning-CommonVoice.ipynb:
```diff
@@ -0,0 +1,666 @@
+    "conf.trainer.precision = 16\n",
+    "\n",
+    "# uncomment if you want to use W&B, you'll need to pass your token at a later step\n",
+    "conf.exp_manager.wandb_logger_kwargs.name = \"common_voice_esperanto\"\n",
```
Add resume flags. It's harmless if it completes the job, but if it fails then at least pointing to same paths with checkpoints will allow continuation

titu1994 commented on tutorials/asr/cloud/aws/SageMaker_ASR_Fine-tuning.ipynb:
```diff
@@ -0,0 +1,419 @@
+   "metadata": {},
+   "source": [
+    "# NeMo ASR Training Using AWS SageMaker"
+   ]
```
Script name is finetuning but it's actually training from scratch?

titu1994 reviewed (changes requested):
Minor changes requested. Very nice tutorials !

SeanNaren commented:
Great review @titu1994 will address. @andrusenkoau for visibility as some of the points will relate to his work too!

titu1994 commented on tutorials/cloud/aws/SageMaker_ASR_Training.ipynb:
```diff
@@ -0,0 +1,418 @@
+    "\n",
+    "1. Setup your AWS Credentials to access SageMaker\n",
+    "2. Download the source code we'll be running\n",
+    "3. Setup AN4 dataset, upload data to S3\n",
```
Isnt AWS a paid resource ? AN4 for that purpose is a bit odd, but I suppose it would cost very little given the size of the dataset. Another option would be to showcase how to load a pretrained data from S3 or something, but that would be too expensive for a tutorial on the nemo side. I think AN4 is fine, just something to note for future consideration.

SeanNaren replied:
In most cases, I assume that people would already have their data on S3 and just omit this honestly. That's fine, just showing someone how easy it is to bring their data to S3 if they haven't already (which I assume is unlikely)!

titu1994 commented on tutorials/cloud/aws/SageMaker_ASR_Training.ipynb:
```diff
@@ -0,0 +1,418 @@
+    "conf.model.n_heads = 4\n",
+    "conf.model.spec_augment.time_masks = 5\n",
+    "\n",
+    "# Set Optimizer parameters\n",
```
Lets note that LR here is a scaler multiplier for Noam, not actually LR of 2.0 (we've had comments before)

SeanNaren replied:
I didn't even know this, thanks for clarification! added.

titu1994 reviewed (approved):
Excellent writeup ! Minor changes only, then ready to merge

SeanNaren commented on nemo/utils/cloud.py:
```diff
@@ -80,3 +82,61 @@ def maybe_download_from_cloud(url, filename, subfolder=None, cache_dir=None, ref
+        return env
+
+    @cluster_environment.setter
+    def cluster_environment(self, env):
```
Agreed, @borda any chance the SageMaker strategy could exist in Lightning? I saw a PR but looks like it was closed: https://github.com/Lightning-AI/lightning/pull/6271

SeanNaren commented on nemo/utils/cloud.py:
```diff
@@ -80,3 +82,61 @@ def maybe_download_from_cloud(url, filename, subfolder=None, cache_dir=None, ref
+    def _install_system_libraries() -> None:
+        os.system('chmod 777 /tmp && apt-get update && apt-get install -y libsndfile1 ffmpeg')
+
+    def _patch_torch_metrics() -> None:
```
@SkafteNicki @borda I had to hack the hash function due to how SageMaker wraps the init of metrics init to do auto-partitioning. In this case, I had to remove the references to self for metrics. Is this something you've ran into before? Have we thought about simplifying the `__hash__` function? I wonder if the extra complexity is even necessary.

SeanNaren commented on nemo/utils/cloud.py:
```diff
@@ -80,3 +82,61 @@ def maybe_download_from_cloud(url, filename, subfolder=None, cache_dir=None, ref
+
+        if dist.get_local_rank():
+            _install_system_libraries()
+        print("WORLD SIZE NEMO", dist.get_world_size())
```
Ah, I removed this. It shouldn't be necessary as Lightning will confirm this later on.

Works fine with a single GPU, `RANK` and `WORLD_SIZE` will not be defined in a single GPU environment, so this code would be skipped.

SeanNaren commented on tutorials/cloud/aws/ASR_Finetuning_at_Scale_with_AWS_SageMaker.ipynb:
```diff
@@ -0,0 +1,751 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# conf.exp_manager.wandb_logger_kwargs.name = \"common_voice_esperanto\"\n",
```
When using multi-node, I only observe the first node/first rank logging. I assume this is correct?

Also I pass the WANDB_AUTH_TOKEN when launching the job as an env variable. This is done at the end of the script when we launch the job using SageMaker (which skips the need to login).

SeanNaren commented on tutorials/cloud/aws/ASR_Finetuning_at_Scale_with_AWS_SageMaker.ipynb:
```diff
@@ -0,0 +1,751 @@
+    "    role=role,\n",
+    "    instance_type=instance_type,\n",
+    "    instance_count=instance_count,\n",
+    "    framework_version=\"1.12.0\",\n",
```
PyTorch 1.13+ isn't supported for SageMaker DDP unfortunately (yet).

SeanNaren commented on tutorials/cloud/aws/ASR_Finetuning_at_Scale_with_AWS_SageMaker.ipynb:
```diff
@@ -0,0 +1,751 @@
+    "  - Make sure you are logged in and then read that agreement.\n",
+    "  - If and only if you agree to the text, then accept the terms.\n",
+    "\n",
+    "Code steps:\n",
```
Done!

SeanNaren commented:
Great review @titu1994, addressed all points. Hopefully the PTL team can shed light if we can move the SageMaker strategy elsewhere. Even in a `lightning-sagemaker` repo which we could ask the user to install would work cc @borda.

titu1994 reviewed (approved):
LGTM, thanks !
